### ISSUE TITLE
Partition split in 5 by RecuperaBit? Most recovered files broken.

### ISSUE THREAD

**Issue by reasv** [Lazza/RecuperaBit#18]
Hi, I'm trying to recover data from a 3TB WD RED drive (WD30EFRX).
The drive itself is healthy and working properly. What happened is that I accidentally wrote an `.iso` to it using `dd`. The iso is 4GB.
This is the original partition table from obtained through `fdisk -l` *before* the accident happened and this table was destroyed:
```
Device          Start        End    Sectors  Size Type
/dev/xvdj1         34     262177     262144  128M BIOS boot
/dev/xvdj2     264192 5848213503 5847949312  2.7T Microsoft basic data
/dev/xvdj3 5848215552 5860532223   12316672  5.9G Linux filesystem
```
This is after the `dd` accident:
```
Device     Boot Start     End Sectors  Size Id Type
/dev/sde1  *        0 8294399 8294400    4G  0 Empty
/dev/sde2         968   63923   62956 30,8M ef EFI (FAT-12/16/32)
```
so the first 4GB of the disk got overwritten.
The MFT was probably completely overwritten by the 4GB iso.
This is why I chose to try your tool after reading the article you wrote about it.

I am interested in recovering the second 2.7tb partition. This is not a windows partition btw, it's just file storage. The `BIOS boot` partition is a residue from an old windows installation and it has no meaning.

Originally there were 83Gb of free space on this partition so this is about 2.6tb of files to recover.

After the analysis these are the `recoverable` partitions I get:
https://pastebin.com/1kWJGuZL
As you can see there are 5 partitions that start at the same point as the original partition (`264192`) but they all end long before the original partition, and all at the same spot (`Offset (b): 135266304`)
```
Partition #248 -> Partition (NTFS, ??? b, 193571 files, Recoverable, Offset: 264192, Offset (b): 135266304, Sec/Clus: 8, MFT offset: 2698650792, MFT mirror offset: None)
Partition #283 -> Partition (NTFS, ??? b, 114301 files, Recoverable, Offset: 264192, Offset (b): 135266304, Sec/Clus: 8, MFT offset: 3418165840, MFT mirror offset: None)
Partition #312 -> Partition (NTFS, ??? b, 391342 files, Recoverable, Offset: 264192, Offset (b): 135266304, Sec/Clus: 8, MFT offset: 4287141256, MFT mirror offset: None)
Partition #451 -> Partition (NTFS, ??? b, 71413 files, Recoverable, Offset: 264192, Offset (b): 135266304, Sec/Clus: 8, MFT offset: 614360032, MFT mirror offset: None)
Partition #1007 -> Partition (NTFS, ??? b, 207562 files, Recoverable, Offset: 264192, Offset (b): 135266304, Sec/Clus: 8, MFT offset: 2592906944, MFT mirror offset: None)
```
I tried recovering all recoverable partitions and these 5 are the only ones that contain files from my original 2.7TB partition. The others are unrelated.
The problem is that the total amount of data recovered from these 5 partitions is about `963GB`, of which more than 100GB are broken (partial) files, and consider that among these broken files there are many that are `0 bytes`, so there is a lot missing.
The data that I could recover is less than 33% overall in GB. Much lower if you only count the % of working files.
Now, I understand that many files are fragmented and that many fragments are probably in the initial 4GB that were overwritten, but to lose almost 70% of the data seems unrealistic.
Also it's weird that the data is divided in 5 partitions starting at the same point.

I did `restore (#) 5` and `restore (#) -1` for all of these partitions.
Most of the files recovered are in `LostFiles` folders. Some are in `Root` but it's a minority.
I meticulously examined each and every `Dir_XXXX` folder (there were thousands), and I sorted them based on content and where the files were originally and I noticed something strange:
Some `Dir_XXXX` folders were actually originally sub-folders of folders recovered from the other partitions.
So for example `Dir_XXXX` in partition #1007 could be the sub-folder of a a folder inside `Dir_YYYY` from partition #248.
From what I understand about your program something is placed inside a `Dir_XXXX` if the parent can't be found. But in these cases the parent was found, just in another partition. These 5 partitions are actually just one.

Another thing I noticed is that files found in a `Dir_XXXX` folder on one partition were originally in the same folder as files I found in a  `Dir_YYYY` folder on another partition. So in these cases the two `Dir_` folder were originally the same folder.

I have also noticed that there are no duplicate files or folders between these 5 partitions. They are not different versions of the same data, I think they are pieces of one partition, but the program thinks they are separate.

This may also be why many files are broken; Perhaps their other pieces are in another partition among these 5.

I am asking because I'm not sure about how to solve this. Is there a way of telling the program to consider all sectors from `264192` to `5848213503` as one partition instead of dividing it in 5?
I think that recovering a lot more of my files should be possible, as only a very small piece of the disk was overwritten.
Perhaps should I use `fdisk` to recreate the partition from `264192` to `5848213503` without formatting it, and then run RecuperaBit again pointing it at this partition? Or is there another solution?
Thank you.

PS: Partition #846 which starts at `1180893537`also contains files from my original partition.
It's the only one that contains my files and doesn't start at `264192` like the other 5.
The reason why I didn't mention it before is that all of the files I extracted from it are broken.
I believe this is because it has a wrong cluster setting (`Sec/Clus: 1` instead of 8).

**Comment by reasv** [comment]
While reading your master thesis on RecuperaBit, I found this part to be relevant (page 49):

> There are two situations in which this approach may lead to an in-correct clustering. The ﬁrst one is that of overlapping partitions with the same MFT offset.
> This was already discussed in Section 4.2, motivating the impossibility of separating ﬁle records in this case. The other one is the presence of a fragmented MFT, which occurs if it becomes bigger than its reserved space because many entries are needed [3].
> If an MFT is fragmented, each part has a different offset, leading to the recognition of multiple ﬁle systems.
> Assuming that the $MFT ﬁle record (id 0) is readable, the pieces may be combined by reading the runlist of its $DATA attribute. In the opposite case, user intervention is required to judge if artifacts are spread across fragments of the same MFT or they belong to different ﬁle systems.

I believe that this situation is case #2, however I'm still not sure on how to solve this.

**Comment by Lazza** [comment]
> I noticed something strange:
Some `Dir_XXXX` folders were actually originally sub-folders of folders recovered from the other partitions.

This is not surprising if those 5 parts belong to the same partition.

> it's weird that the data is divided in 5 partitions starting at the same point

It isn't weird at all. If the MFT is fragmented and the $MFT file record is missing, there is no reliable way to _automatically_ guess that those are parts of the same partition.

> These 5 partitions are actually just one.

Yes I understand your point and I trust this, I just want to underline the fact that RecuperaBit does not attempt brave guesses to merge partitions because that could lead to unreliable results. A forensic tool surely should not try to "invent" information that cannot be deduced from the disk image.

> Is there a way of telling the program to consider all sectors from `264192` to `5848213503` as one partition instead of dividing it in 5?

No, but the idea of adding a feature to **manually** merge two partitions (through another command) makes sense. Can you open an issue for this?

> Perhaps should I use `fdisk`

Partition tables are not used anyways because they can contain *anything* and cannot be trusted. Also, you would still be missing the file record for `$MFT`.

---

> I believe that this situation is case #2

The second case would only work if the file record for `$MFT` is available, but you overwrote it.

---

By the way it's a bit weird that you are missing `$MFTMirr` as well. Isn't it in any of those 5 partitions? If the $MFTMirr is pristine then this is a totally different issue and we should investigate it further.

**Comment by reasv** [comment]
>but the idea of adding a feature to manually merge two partitions (through another command) makes sense. Can you open an issue for this?
Sure.

>By the way it's a bit weird that you are missing $MFTMirr as well. Isn't it in any of those 5 partitions? If the $MFTMirr is pristine then this is a totally different issue and we should investigate it further.
If you mean in the root directory, no. It's just not there in any of the partitions that contain my files.
I don't think it's in the lost files because I've examined all of them.

There are some $MFTMirr files in the other partitions that seem unrelated to my files though.
But of these 5 partitions, none of them have any MFTMirr file.

By the way, do you think that merging the partitions could do something for at least some of the many files that came out damaged?

**Comment by Lazza** [comment]
> It's just not there in any of the partitions that contain my files

OK, so you need to merge manually.

I have closed #19 so try that one.

> do you think that merging the partitions could do something for at least some of the many files that came out damaged?

Most likely, yes. 😉
